chore: added linea state bridge and documented them

KodeSage opened the pull request:
Linked to this issue #1 
- Added The `IMessageService.sol` Inteface: The message service is responsible for cross-chain messages between Ethereum and Linea.
- Added The `LineaWorldID.sol` interface
- Added the 'LineaStateBridge.sol` Contract.

kfastov commented:
@KodeSage Looks lgtm for me, I'll fix CI in the main branch and look more deeply tomorrow

KodeSage commented:
alright @kfastov Thank you.

kfastov commented:
@KodeSage Can you please run `forge fmt` to lint the source automatically?

KodeSage commented:
alright done! @kfastov

kfastov commented:
@KodeSage What about `remappings.txt`? I saw you deleted it, did it cause any problem?

KodeSage commented:
@kfastov The `remappings.txt` was flaggling an error when I tried building it locally, that's why I removed it. I will use this `bun` instruction to try and make it build now.

KodeSage commented:
Hello @kfastov , I just updated necessary dependencies and added back the `remappings.txt` and the build was successful.
<img width="520" alt="Screenshot 2024-06-26 at 22 01 47" src="https://github.com/kfastov/worldcoin-bridge-linea/assets/40062598/fa196beb-a122-47cc-bfde-6b667c611c9a">

0xprinc commented:
Hey, @KodeSage @kfastov I have used modifiers of `MessageServiceBase` as mentioned in Linea Docs and not of the older `CrossDomainOwnable3` contract.
Now there are two authorized addresses `messageService` and `remoteSender` and not a single owner(previous implementation).
So instead of calling 
```
ICrossDomainOwnable3.transferOwnership, (_owner, _isLocal)
```

call this in you implemented function `transferOwnershipLineas`

```
MessageServiceBase.updateMessageServiceBase(address _messageService, address _remoteSender)
```

KodeSage commented:
@kfastov I just pulled from the main branch and resolved neccssary conflict.

kfastov commented:
@KodeSage You are great. I was going to ask about pulling the latest main commits, but you did it so fast

KodeSage commented:
> @KodeSage You are great. I was going to ask about pulling the latest main commits, but you did it so fast

Thank you 😊 @kfastov

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+    /// @param messageService The new owner of the LineaWorldID contract
+    /// @param remoteAddress Whether the ownership transfer is local (Linea/Linea Stack chain EOA/contract)
+    /// or an Ethereum EOA or contract
+    event OwnershipTransferredLinea(
+        address indexed previousOwner, address indexed messageService, address remoteAddress
+    );
```
I guess it's not "OwnershipTransferred" anymore, because we talk about message service address change. Since we now have 2 moving parts: "message service" and "authorized remote sender", we may want 2 events for registering changes of those entities, and maybe separate methods to modify them

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+    /// @param _crossDomainMessenger L1CrossDomainMessenger contract used to communicate with the desired Linea
+    /// Stack network
+    /// @custom:revert if any of the constructor params addresses are zero
+    constructor(address _worldIDIdentityManager, address _lineaWorldIDAddress, address _crossDomainMessenger) {
```
> _crossDomainMessenger
I think if we use Linea terminology, we may name it "messageService", correct me if I'm wrong

KodeSage replied:
alright, I think this would work, will make changes to it now to `messageService`

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+        uint256 latestRoot = IWorldIDIdentityManager(worldIDAddress).latestRoot();
+
+        // The `encodeCall` function is strongly typed, so this checks that we are passing the
+        // correct data to the optimism bridge.
```
Seems like a leftover from an OP state bridge
```suggestion
        // correct data to the Linea messaging service.
```

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+        emit RootPropagated(latestRoot);
+    }
+
+    /// @notice Adds functionality to the StateBridge to transfer ownership
+    /// of LineaWorldID to another contract on L1 or to a local Linea Stack chain EOA
+    /// @dev  _messageService should be hardcoded to 0x508Ca82Df566dCD1B0DE8296e70a96332cD644ec for L2 Linea
+    /// @dev _remoteSender should be address of L1 LineaStateBridge
+    /// @custom:revert if _owner is set to the zero address
+    // function transferOwnershipLineas(address _messageService, address _remoteSender) external onlyOwner {
+    //     if (_remoteSender == address(0) || _messageService == address(0)) {
+    //         revert AddressZero();
+    //     }
+
+    //     // The `encodeCall` function is strongly typed, so this checks that we are passing the
+    //     // correct data to the Linea Stack chain bridge.
+    //     bytes memory message =
+    //         abi.encodeCall(MessageServiceBase.updateMessageServiceBase(_messageService, _remoteSender));
+
+    //     IMessageService(crossDomainMessengerAddress).sendMessage(
+    //         // Contract address on the Linea Stack Chain
+    //         lineaWorldIDAddress,
+    //         _gasLimitTransferOwnership,
+    //         message
+    //     );
+
+    //     emit OwnershipTransferredLinea(owner(), _messageService, _remoteSender);
+    // }
```
There is a commented out implementation of `transferOwnershipLineas`. Please remove it if it doesn't hold any important information

KodeSage replied:
alright, will do just that!

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+
+    //     emit OwnershipTransferredLinea(owner(), _messageService, _remoteSender);
+    // }
+    function transferOwnershipLineas(address _messageService, address _remoteSender) external onlyOwner {
```
Could you please explain the naming? Does the last `s` have any special meaning?

KodeSage replied:
nope there isn't any special meaning, will remove the `s`

kfastov commented on src/LineaStateBridge.sol:
```diff
@@ -1,17 +1,251 @@
+    /// @param rootHistoryExpiry The new root history expiry
+    event SetRootHistoryExpiry(uint256 rootHistoryExpiry);
+
+    /// @notice Emitted when the StateBridge sets the gas limit for sendRootOp
+    /// @param _lineaGasLimit The new _lineaGasLimit  for sendRootOp
+    event SetGasLimitPropagateRoot(uint32 _lineaGasLimit);
+
+    /// @notice Emitted when the StateBridge sets the gas limit for SetRootHistoryExpiry
+    /// @param _lineaGasLimit The new _lineaGasLimit  for SetRootHistoryExpiry
+    event SetGasLimitSetRootHistoryExpiry(uint32 _lineaGasLimit);
+
+    /// @notice Emitted when the StateBridge sets the gas limit for transferOwnershipOp
+    /// @param _lineaGasLimit The new _lineaGasLimit  for transferOwnershipOptimism
+    event SetGasLimitTransferOwnershipLinea(uint32 _lineaGasLimit);
```
Comments still mention Optimism chain, I guess it was not intended

kfastov replied:
Sorry, left you a bit of a work. If it's too late, I think we can finish it tomorrow

kfastov reviewed (changes requested):
Looks good, added some comments, mostly about naming and leftover comments

kfastov commented on src/interfaces/ILineaWorldID.sol:
```diff
@@ -2,7 +2,7 @@
 /// @title Interface for the LineaWorldID contract
-/// @author Worldcoin and 0xprinc
+/// @author Worldcoin and 0xprinc x James Harrison.
 /// @notice Interface for the CrossDomainOwnable contract for the Linea L2
```
Since we don't have CrossDomainOwnable implementation for Linea (yet?) and use raw MessageService contract, let's call it ... MessageServiceBase contract (because it inherits from `MessageServiceBase`? Or maybe you have better options

kfastov reviewed (changes requested):
I highlighted minor naming issues with suggestions how to fix them.
- I want to eliminate CrossDomainMessenger mentions, because it's an OP Stack entity
- In some places comments are not consistent with the code

kfastov commented:
@KodeSage I hope it's the last one 😅

KodeSage commented:
> @KodeSage I hope it's the last one 😅

I hope so too @kfastov 😩, I have made the changes.

kfastov commented:
Cool, I think we can merge it now! 🎉
